src/ui/api/metadata.go

```diff
@@ -0,0 +1,245 @@
+type MetadataAPI struct {
+	BaseController
+	metaMgr metamgr.ProjectMetadataManager
+	project *models.Project
+	name    string
+}
+
+// Prepare ...
+func (m *MetadataAPI) Prepare() {
+	m.BaseController.Prepare()
+
+	m.metaMgr = m.ProjectMgr.GetMetadataManager()
+
+	// the project manager doesn't use a project metadata manager
+	if m.metaMgr == nil {
+		log.Debug("the project manager doesn't use a project metadata manager")
+		m.RenderError(http.StatusMethodNotAllowed, "")
+		return
+	}
+
+	id, err := m.GetInt64FromPath(":id")
+	if err != nil || id <= 0 {
+		text := "invalid project ID: "
+		if err != nil {
+			text += err.Error()
+		} else {
+			text += fmt.Sprintf("%d", id)
+		}
+		m.HandleBadRequest(text)
+		return
+	}
+
+	project, err := m.ProjectMgr.Get(id)
+	if err != nil {
+		m.ParseAndHandleError(fmt.Sprintf("failed to get project %d", id), err)
+		return
+	}
+
+	if project == nil {
+		m.HandleNotFound(fmt.Sprintf("project %d not found", id))
+		return
+	}
+
+	m.project = project
+
+	switch m.Ctx.Request.Method {
+	case http.MethodGet:
+		if !project.IsPublic() {
+			if !m.SecurityCtx.IsAuthenticated() {
+				m.HandleUnauthorized()
+				return
+			}
+			if !m.SecurityCtx.HasReadPerm(project.ProjectID) {
+				m.HandleForbidden(m.SecurityCtx.GetUsername())
+				return
+			}
+		}
+	case http.MethodPost, http.MethodPut, http.MethodDelete:
+		if !m.SecurityCtx.IsAuthenticated() {
+			m.HandleUnauthorized()
+			return
+		}
+		if !m.SecurityCtx.HasAllPerm(project.ProjectID) {
+			m.HandleForbidden(m.SecurityCtx.GetUsername())
+			return
+		}
+	default:
+		log.Debugf("%s method not allowed", m.Ctx.Request.Method)
+		m.RenderError(http.StatusMethodNotAllowed, "")
+		return
+	}
+
+	name := m.GetStringFromPath(":name")
+	if len(name) > 0 {
+		m.name = name
+		metas, err := m.metaMgr.Get(project.ProjectID, name)
+		if err != nil {
+			m.HandleInternalServerError(fmt.Sprintf("failed to get metadata of project %d: %v", project.ProjectID, err))
+			return
+		}
+		if len(metas) == 0 {
+			m.HandleNotFound(fmt.Sprintf("metadata %s of project %d not found", name, project.ProjectID))
+			return
+		}
+	}
+}
+
+// Get ...
+func (m *MetadataAPI) Get() {
+	var metas map[string]string
+	var err error
+	if len(m.name) > 0 {
+		metas, err = m.metaMgr.Get(m.project.ProjectID, m.name)
+	} else {
+		metas, err = m.metaMgr.Get(m.project.ProjectID)
+	}
+
+	if err != nil {
+		m.HandleInternalServerError(fmt.Sprintf("failed to get metadata %s of project %d: %v", m.name, m.project.ProjectID, err))
+		return
+	}
+	m.Data["json"] = metas
+	m.ServeJSON()
+}
+
+// Post ...
+func (m *MetadataAPI) Post() {
+	var metas map[string]string
+	m.DecodeJSONReq(&metas)
+
+	ms, err := validateProjectMetadata(metas)
+	if err != nil {
+		m.HandleBadRequest(err.Error())
+		return
+	}
+
+	if len(ms) != 1 {
+		m.HandleBadRequest("invalid request: has no valid key/value pairs or has more than one valid key/value pairs")
+		return
+	}
+
+	name := ""
+	for name = range metas {
```

Comment:
keys := reflect.ValueOf(a).MapKeys()
    strkeys := make([]string, len(keys))
    for i := 0; i < len(keys); i++ {
        strkeys[i] = keys[i].String()
    }
    fmt.Print(strings.Join(strkeys, ","))